lib/Dialect/Ibis/Transforms/IbisContainerize.cpp

```diff
@@ -0,0 +1,134 @@
+//===- IbisContainerize.cpp - Implementation of containerizing ------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+
+#include "circt/Dialect/Ibis/IbisDialect.h"
+#include "circt/Dialect/Ibis/IbisOps.h"
+#include "circt/Dialect/Ibis/IbisPasses.h"
+#include "circt/Dialect/Ibis/IbisTypes.h"
+
+#include "circt/Support/SymCache.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace circt;
+using namespace ibis;
+
+// Iterates the symbolcache until a unique name is found.
+static StringAttr getUniqueName(mlir::MLIRContext *ctx, StringAttr baseName,
+                                SymbolCache &symCache) {
+  StringAttr uniqueName = baseName;
+  int uniqueCntr = 0;
+  while (symCache.getDefinition(uniqueName))
+    uniqueName =
+        StringAttr::get(ctx, baseName.strref() + "_" + Twine(uniqueCntr++));
+  return uniqueName;
+}
+namespace {
+
+struct OutlineContainerPattern : public OpConversionPattern<ContainerOp> {
+  OutlineContainerPattern(MLIRContext *context, SymbolCache *symCache)
+      : OpConversionPattern<ContainerOp>(context), symCache(symCache) {}
+
+  using OpAdaptor = typename OpConversionPattern<ContainerOp>::OpAdaptor;
+
+  LogicalResult
+  matchAndRewrite(ContainerOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Outline the container into the module scope, by prefixing it with the
+    // parent class name.
+    auto parentClass =
+        dyn_cast_or_null<ClassOp>(op.getOperation()->getParentOp());
+    if (!parentClass)
+      return failure();
```

Comment:
IIRC there's a way to explain why this wasn't a match which would help debug.

Reply:
Modifiying it to an assert - given the `!isa` dynamic legality change, the parent should never be anything but a classop.